### DISCUSSION TITLE
Create a chart for cumulative delta time

### DISCUSSION THREAD

**Discussion by WarmBed** [theOehrly/Fast-F1#141] [Q&A]
Hi there,

I was trying to create a chart for cumulative delta time for each lap and every driver. 

I can plot the lap time delta for each driver but I can't cumulative them. 

Here is my code. 

Thank you 

![擷取](https://user-images.githubusercontent.com/72774359/161304397-7a69d79a-d01e-40f5-9cc8-fe6944c991a3.PNG)

```
# %%
from cProfile import label
from re import X
import matplotlib.font_manager as fm
import fastf1.plotting
from matplotlib.collections import LineCollection
from matplotlib import cm
import numpy as np
import pandas as pd
from timple.timedelta import strftimedelta
from fastf1.core import Laps
import fastf1 as ff1
from fastf1 import plotting
from fastf1 import utils
from matplotlib import pyplot as plt
plotting.setup_mpl()
fastf1.Cache.enable_cache('doc_cache')  
fastf1.plotting.setup_mpl()

# %%
der1 = 'VER'
der2 = 'PER'

year = 2022
wknd = 2
ses = 'R'

der1_LapTime=[]
der2_LapTime=[]

der1_seleLap = 21
der2_seleLap = 18

session = fastf1.get_session(year, wknd, ses)
session.load()


# %%
driver_everylap={}
for driver in list(session.results['Abbreviation']):
  driver_everylap[driver] = []
  for laptime in range(len(list(session.laps.pick_driver(driver)['LapTime']))):
      driver_everylap[driver].append(list(session.laps.pick_driver(driver)['LapTime'])[laptime].total_seconds())

all_everylap=[]
for driver in list(session.results['Abbreviation']):
    all_everylap.append(np.nanmean(driver_everylap[driver]))
    ave_laptime=np.nanmean(all_everylap)

# %%

plt.figure(figsize=(10,5), dpi=100)
for driver_ID in list(session.results['Abbreviation']):
    lap_data_list = driver_everylap[driver_ID]
    accumulated_delta=[]
    for i in range(0,len(driver_everylap[driver_ID])):
        accumulated_delta.append(lap_data_list[i]-lap_data_list[i-1])
        total_delta =[0]
        for i in range(1,len(accumulated_delta)):
            total_delta.append(total_delta[i-1]+accumulated_delta[i])
    plt.plot(total_delta,label=driver_ID)
    #plt.plot(accumulated_delta,label=driver_ID)



plt.legend()
plt.show()


```

**Comment by theOehrly** [comment]
There are two problems. First, calculating the delta creates some NaN values. NaN + anything else = NaN, therefore all values of your total delta are NaN. Second, one of your nested for loops is indented incorrectly, imo. Compare my changes:

```
plt.figure(figsize=(10, 5), dpi=100)
for driver_ID in list(session.results['Abbreviation']):
    lap_data_list = driver_everylap[driver_ID]
    accumulated_delta = []
    for i in range(0, len(driver_everylap[driver_ID])):
        delta = lap_data_list[i] - lap_data_list[i - 1]
        delta = 0 if np.isnan(delta) else delta
        accumulated_delta.append(delta)
    total_delta = [0]
    for i in range(1, len(accumulated_delta)):
        total_delta.append(total_delta[i - 1] + accumulated_delta[i])
    plt.plot(total_delta, label=driver_ID)
    # plt.plot(accumulated_delta, label=driver_ID)
```

![grafik](https://user-images.githubusercontent.com/23384863/161439878-bab61d9c-698d-438b-92b5-0d94353ab67e.png)
Is this what you were trying to create?

And here's a more "pandas" way of doing the same thing:

```
import fastf1.plotting
import pandas as pd
from fastf1 import plotting
from matplotlib import pyplot as plt

plotting.setup_mpl()
fastf1.Cache.enable_cache('tmp_cache')
fastf1.plotting.setup_mpl()

der1 = 'VER'
der2 = 'PER'

year = 2022
wknd = 2
ses = 'R'

session = fastf1.get_session(year, wknd, ses)
session.load()

plt.figure(figsize=(10, 5), dpi=100),

for driver_abb in session.results['Abbreviation']:
    laps = session.laps.pick_driver(driver_abb).reset_index(drop=True)

    laps['LapTimeDelta'] = pd.Timedelta(0)
    laps['TotalLapTimeDelta'] = pd.Timedelta(0)

    laps.loc[1:, 'LapTimeDelta'] = laps['LapTime'].iloc[1:].to_numpy() - laps['LapTime'].iloc[:-1].to_numpy()
    laps['LapTimeDelta'] = laps['LapTimeDelta'].fillna(pd.Timedelta(0))
    laps['TotalLapTimeDelta'] = laps['LapTimeDelta'].cumsum()

    plt.plot(laps['TotalLapTimeDelta'].dt.total_seconds(), label=driver_abb)


plt.legend()
plt.show()
```

**Reply by WarmBed** [reply]
Hi Oehrly, 

Thanks for your reply. I realized the issues now. I will prefer to use the first one you give me. 

I will try it later.

Thanks!

What I try to  create is like this picture. But the firt step is I need to create the delta time and total delta time. 


![1_4yok31MCV1ma8IzMUbQ9wA](https://user-images.githubusercontent.com/72774359/161445536-ff52fff4-e84a-4ab4-9d85-1ea13b9f8941.png)

**Reply by markste-in** [reply]
I am kinda unsure what you mean by "cumulative delta time". Is this the "sum of the lap time delta to the previous round (driver by driver)" or is it the "sum of the deltas to the previous round of the fastest driver" or is it the "sum of the delta to fastest lap time seen" ... or idk..

What do you want as "reference time"?
Tho I am pretty sure it can be done in a few lines of code

**Reply by WarmBed** [reply]
> 

Hi, markste,

"sum of the lap time delta to the previous round (driver by driver) " This is what I actually wanted. 

The reference time is the ave time in total laps. 

As you saw, I already calculation the ave_laptime=np.nanmean(all_everylap) This will be my baseline.

So using this chart I can see who is trying to push the car or which lap has fight(maybe). 

Thanks!
